### ISSUE TITLE
Transpose: give the function empty list and get the error : Length of the 'dim' parameter must be equal to 4

### ISSUE THREAD

**Issue by FrozenGene** [tf-coreml/tf-coreml#210]
When I convert shuffleseg tensorflow model, I meet the following error messgae:

  File "/usr/local/lib/python3.6/site-packages/tfcoreml-0.2.0-py3.6.egg/tfcoreml/_tf_coreml_converter.py", line 552, in convert
  File "/usr/local/lib/python3.6/site-packages/tfcoreml-0.2.0-py3.6.egg/tfcoreml/_tf_coreml_converter.py", line 315, in _convert_pb_to_mlmodel
  File "/usr/local/lib/python3.6/site-packages/tfcoreml-0.2.0-py3.6.egg/tfcoreml/_ops_to_layers.py", line 176, in convert_ops_to_layers
  File "/usr/local/lib/python3.6/site-packages/tfcoreml-0.2.0-py3.6.egg/tfcoreml/_layers.py", line 877, in transpose
  File "/usr/local/lib/python3.6/site-packages/coremltools/models/neural_network.py", line 2044, in add_permute
    **raise ValueError("Length of the 'dim' parameter must be equal to 4")**

I check the code of def transpose.
```python
context.builder.add_permute(output_name, [], input_name, output_name)
```
But the coremltools required:
```python
 if len(dim) != 4:
       raise ValueError("Length of the 'dim' parameter must be equal to 4")
```

we give empty list. Could we fix it? Thanks

P.S.
In my case, give it [0,3,2,1] can work fine.

**Comment by aseemw** [comment]
I can put up a fix. Can you please provide a tiny test network that reproduces this error? Thanks.

**Comment by FrozenGene** [comment]
[shuffleseg_optimized_graph_v11.pb.zip](https://github.com/tf-coreml/tf-coreml/files/2153697/shuffleseg_optimized_graph_v11.pb.zip)

You can translate this model and reproduce the error. The information maybe you need:

input_layer:                                network/input/input 
input_layer_shape:                    1,256,512,3
output_layer:                              network/output/final_result

**Comment by nwestlake** [comment]
I believe this is the correct fix. Please could you confirm. 

```
diff --git a/tfcoreml/_layers.py b/tfcoreml/_layers.py
index b1c4ccc..0987203 100644
--- a/tfcoreml/_layers.py
+++ b/tfcoreml/_layers.py
@@ -937,7 +937,7 @@ def transpose(op, context):
   coreml_axes[target_width_idx] = 3
   coreml_axes[target_channel_idx] = 1
 
-  context.builder.add_permute(output_name, [], input_name, output_name)
+  context.builder.add_permute(output_name, coreml_axes, input_name, output_name)
   context.translated[output_name] = True
 
 def real_div(op, context):

```

**Comment by nwestlake** [comment]
Apologies, we did some more testing and it seems it isn't the correct fix but #249 is.

We have tested it and found it to work with swapping but not a total permutation. I believe this but may be in coreml as I cannot locate it in tf-coreml or coremltools. 

The test network we used is as follows:

    import numpy as np
    import os
    import tensorflow as tf
    import tfcoreml

    from tensorflow.python.tools.freeze_graph import freeze_graph

    def makedirs_if_needed(dir):
        if not os.path.exists(dir):
            os.makedirs(dir)


    export_dir = "test_transpose"

    graph_filename = 'graph.pbtxt'
    frozen_graph_filename = 'graph.pb'
    mlmodel_filename = "model.mlmodel"
    checkpoint_path = os.path.join(export_dir, "checkpoint")

    makedirs_if_needed(export_dir)

    graph = tf.Graph()
    with graph.as_default():

        input_ph = tf.placeholder(
            tf.float32,
            [1, 2, 3, 4],
            name='input')

        var = tf.get_variable("dummy", shape=[1])

        output = input_ph

        #These work
        output = tf.transpose(output, perm=[0, 2, 1, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 2, 1, 3])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 2, 1, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 2, 1, 3])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])
        output = tf.transpose(output, perm=[0, 1, 2, 3])
        output = tf.transpose(output, perm=[0, 3, 2, 1])
        output = tf.transpose(output, perm=[0, 1, 3, 2])

        #But these do not
        #output = tf.transpose(output, perm=[0, 2, 3, 1])
        #output = tf.transpose(output, perm=[0, 3, 1, 2])


        with tf.Session() as sess:
            saver = tf.train.Saver()
            tf.global_variables_initializer().run()
            _ = sess.run(output, feed_dict={input_ph: np.random.rand(1, 2, 3, 4)})
            input_checkpoint = saver.save(
                sess, os.path.join(checkpoint_path, "latest"))
            graph_path = os.path.join(export_dir, graph_filename)
            frozen_graph_path = os.path.join(export_dir, frozen_graph_filename)
            mlmodel_path = os.path.join(export_dir, mlmodel_filename)

            tf.train.write_graph(sess.graph_def, export_dir, graph_filename)
            freeze_graph(input_graph=graph_path,
                         input_saver="",
                         input_binary=False,
                         input_checkpoint=input_checkpoint,
                         output_node_names=output.name.split(":")[0],
                         restore_op_name=None,
                         filename_tensor_name=None,
                         output_graph=frozen_graph_path,
                         clear_devices=True,
                         initializer_nodes="")

    tfcoreml.convert(
        tf_model_path=frozen_graph_path,
        mlmodel_path=mlmodel_path,
        output_feature_names=[output.name])


and then running with
    import coremltools
    import numpy as np
    import os
    import tensorflow as tf

    TEST_DIR = "test_transpose"


    def predictViaSavedModel(model_path, data):
        with tf.Session(graph=tf.Graph()) as sess:

            with tf.gfile.GFile(os.path.abspath(model_path), "rb") as f:
                graph_def = tf.GraphDef()
                graph_def.ParseFromString(f.read())

            tf.import_graph_def(graph_def, name='')

            feed_dict = {"inputs0": data}

            result = tf.get_default_graph().get_tensor_by_name("Reshape:0")

            return sess.run(result, feed_dict=feed_dict)


    def predictViaCoreML(model_path, data):
        feed_dict = {
            "audio_inputs__0": data.tolist(),
        }

        model = coremltools.models.MLModel(model_path)

        print(model.input_description)
        print(model.output_description)

        output = model.predict(feed_dict, useCPUOnly=True)

        return np.squeeze(
                  np.array(output['Tranpose__ '], #Put number in here
                           dtype=np.float64)
               )
